### DISCUSSION TITLE
Deserialiaze with null vs constructor with 0 parameter

### DISCUSSION THREAD

**Discussion by jmfayard** [charleskorn/kaml#251] [General]
Hello

I try to model Github Actions workflow like this

```kotlin
data class YamlWorkflow(
    val name: String,
    val on: YamlWorkflowTriggers,
)
data class YamlWorkflowTriggers(
    val push: Push? = null,
    val pull_request: PullRequest? = null,
)
```

Deserialization works if I pass parameters to `push:` and `pull_request:`

But GitHub Actions also allows to say:

```yaml
name: all-triggers

on:
  push:
  pull_request:
```

Here I expect: `YamlWorkflowTriggers(Push(), PullRequest())`

But I get `YamlWorkflowTriggers(null, null)`

as if the YAML file was

```yaml
name: all-triggers

on:
```

**Comment by charleskorn** [comment]
Thanks for the question @jmfayard. I'm guessing GitHub Actions is doing some magic under the hood there - in YAML, an empty value like that is considered a null value. 

One solution might be to write a custom serializer for `YamlWorkflowTriggers` that translates a present-but-null value like that into an empty object.

**Reply by jmfayard** [reply]
They even accept this syntax :(

```yaml
on: [ push, pull_request ]
```

**Reply by charleskorn** [reply]
That's not going to map nicely to that Kotlin model either - in order to support all of these different scenarios, you'll need to create a custom serializer unfortunately.

**Reply by jmfayard** [reply]
Instead of fighting the type system, I ended up writing an intermediate function `fun String.normalizeYaml(): String `
 that normalize the input to

```yaml
name: all-triggers

on:
  push: {}
  pull_request: {}
```
